hibernate-core/src/main/java/org/hibernate/engine/internal/TwoPhaseLoad.java

```diff
@@ -206,62 +206,91 @@ public static void initializeEntityEntryLoadedState(
+					// hydratedState[i] needs to remain LazyPropertyInitializer.UNFETCHED_PROPERTY so that
+					// setPropertyValues() below (ultimately AbstractEntityTuplizer#setPropertyValues) works properly
+					// No resolution is necessary, unless the lazy property is a collection.
+					if ( types[i].isCollectionType() ) {
+						// IMPLEMENTATION NOTE: this is a lazy collection property on a bytecode-enhanced entity.
+						// HHH-10989: We need to resolve the collection so that a CollectionReference is added to StatefulPersistentContext.
+						// As mentioned above, hydratedState[i] needs to remain LazyPropertyInitializer.UNFETCHED_PROPERTY
+						// so do not assign the resolved, uninitialized PersistentCollection back to hydratedState[i].
+						Boolean overridingEager = getOverridingEager(
+								session,
+								entityName,
+								propertyNames[i],
+								types[i],
+								debugEnabled
+						);
+						types[i].resolve( value, session, entity, overridingEager );
+					}
 				}
+				else if ( value != PropertyAccessStrategyBackRefImpl.UNKNOWN ) {
+					if ( debugEnabled ) {
+						final boolean isLazyEnhanced = persister.getBytecodeEnhancementMetadata()
+								.getLazyAttributesMetadata()
+								.getLazyAttributeNames()
+								.contains( propertyNames[i] );
+						LOG.debugf(
+								"Attribute (`%s`)  - enhanced for lazy-loading? - %s",
+								propertyNames[i],
+								isLazyEnhanced
+						);
+					}
 
-				// we know value != LazyPropertyInitializer.UNFETCHED_PROPERTY
-				Boolean overridingEager = getOverridingEager( session, entityName, propertyNames[i], types[i], debugEnabled );
-				hydratedState[i] = types[i].isEntityType()
-						? entityResolver.resolve( (EntityType) types[i], value, session, entity, overridingEager )
-						: types[i].resolve( value, session, entity, overridingEager );
-			}
-			else {
-				if ( debugEnabled ) {
-					LOG.debugf( "Skipping <unknown> attribute : `%s`", propertyNames[i] );
+					// we know value != LazyPropertyInitializer.UNFETCHED_PROPERTY
+					Boolean overridingEager = getOverridingEager(
+							session,
+							entityName,
+							propertyNames[i],
+							types[i],
+							debugEnabled
+					);
+					hydratedState[i] = types[i].isEntityType()
+							? entityResolver.resolve( (EntityType) types[i], value, session, entity, overridingEager )
+							: types[i].resolve( value, session, entity, overridingEager );
+				}
+				else {
+					if ( debugEnabled ) {
+						LOG.debugf( "Skipping <unknown> attribute : `%s`", propertyNames[i] );
+					}
 				}
-			}
 
-			if ( session.getFetchGraphLoadContext() != fetchGraphContext ) {
 				session.setFetchGraphLoadContext( fetchGraphContext );
 			}
 		}
+		finally {
+			// HHH-14097
+			// Fetch entity graph should be applied only once on top level (for root hydrated object)
+			// e.g., see org.hibernate.loader.Loader for details
+			session.setFetchGraphLoadContext( null );
```

Comment:
I'm not sure about this, are we entirely sure that the fetch graph is no longer needed at this point?
Wouldn't it be possible for the fetch grapth to need to span multiple load events?

IF this is correct, should line 285 be removed as it seems redundant?

Reply:
For non-toplevel traversal, Line #285 will take effect to backtrack to upper level after the above statement. The above statement will impact only top level (I made it explicitly by including "on top level" in the comments above). I admit the code change is difficult to understand for basically 'fetch graph' implementing is sort of based on hack (existing v5 infrastructure makes it difficult to implement elegantly), but v6 based on path retrieval makes the implementation much easier and more elegant. After this PR gets merged, I'll port it into v6 to see whether the issue still remains.

Reply:
I doubled checked and found the backtracking is working as expected (without it our graph testing cases would fail). Could you take a new look at this PR? Thanks.